### DISCUSSION TITLE
Unable to detect left/right click and long press (e.g. for updating video progress)

### DISCUSSION THREAD

**Discussion by andremaximino-wit** [react-native-tvos/react-native-tvos#332] [Feature Requests]
## Description
Hello,
I'm trying to catch the Click Left or Right event and the Long Press Left or Right event on the Apple TV remote, in order to update the progress bar of a custom video player. All I can do until now is catch the tap event or the click event to select something.

Is there some way to listen to these events?

## React Native version:
4.14.0

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. Click left on a touchable item
2.

## Expected Results
Left click event should be detected.

**Comment by douglowder** [comment]
These events are not yet supported in `TVEventHandler`.  I'll take a look to see how much work it will be to add them.

**Comment by andremaximino-wit** [comment]
Thank you. 
I will wait for your input about this.

**Comment by andremaximino-wit** [comment]
Hi!
Do you have any new input on this? 

Thanks

**Comment by douglowder** [comment]
I have looked into this recently, and it appears to require a custom gesture recognizer for the select gesture, plus pulling in GameController code so you can know where the user is touching the Siri touchpad.  I'm following an idea similar to https://dcordero.medium.com/directional-clicks-on-tvos-c711a2faf71a .  I'll post here as I make progress.

**Comment by andrefbmaximino** [comment]
Thanks for looking into this,
I'll wait for your feedback.

**Comment by douglowder** [comment]
@andrefbmaximino just to keep you updated... I finally did get this to work over the past couple of days -- once the code is cleaned up better, I'll create a feature branch and you can try it out.

**Comment by andremaximino-wit** [comment]
Thanks for the update @dlowder-salesforce ! I will wait for the feature branch. Cheers

**Comment by douglowder** [comment]
Ok @andrefbmaximino the feature branch is up -- https://github.com/react-native-tvos/react-native-tvos/tree/left-right-press

Please try this out and let me know what you think.

For an existing project, you can modify your `package.json` by modifying the react-native dependency:

```json
"dependencies": {
    "react": "16.13.1",
    "react-native": "https://github.com/react-native-tvos/react-native-tvos#left-right-press"
  },
```

Then pick up the changes:

- Remove `node_modules` and run `yarn`
- Change to `ios` directory, remove `Pods` and `Podfile.lock`, and run `pod install`

You can also start a new project if you wish:

```bash
$ react-native init TestLeftRightPress --template=https://github.com/react-native-tvos/react-native-tvos#left-right-press
```


The implementation uses `GameController` tvOS framework, so you will have to build and test on a real device.  You should see `stepLeft` and `stepRight` events from the TVEventHandler when the user presses the upper right and left corners of the Siri remote.

The new project template has changes to `App.js` that show the events on the screen as a demo.

At the moment, the decision of when the user has pressed an upper corner is hardcoded in `RCTTVView.m`.  We can discuss how best to make that adjustable if that turns out to be useful.  Also, I'm flexible on the naming of the events.... we could call them `stepForward` and `stepBackward`, for example.

**Comment by andremaximino-wit** [comment]
Hi @dlowder-salesforce 

I've checked the code you've provided and it works on our side. We will try to adapt and use these events in our project. The naming of the events seems fine for us.
Do you know if it's possible to detect the event a little bit down in the remote, as it is only detecting right in the corner. If I press the right edge of the remote in the middle, it doesn't detect this event. Is it something that we can change?
Thank you for all the help with this issue. 
We will get back to you if further questions surge.

**Comment by douglowder** [comment]
Great that this is working! If the behavior is not quite right, you can change the logic or threshold in the `handleSelect:` method and the helper methods it calls to determine what kind of event to fire. If you do find something that works better, please send a PR 😊

**Comment by NickFoden** [comment]
This saved my bacon today, as i had expected to find "right", "left" etc when swiping as `eventType` of events when listening in `  useTVEventHandler(myTVEventHandler);` and only had "blur" "select" "focus" etc and no directional feedback.  (I am developing only in simulator so far this week)

Will this feature branch be raised as pr/merged to main in near future? 

Big fan of the work here @douglowder and if i can help test or with docs or anything please let me know.

**Comment by douglowder** [comment]
@NickFoden the feature branch I created for this is not for left and right swipes, but for pressing the upper left and right corners of the Siri remote, as when stepping forward or backward on a video.  You should already be seeing "left" and "right" event types if you swipe from one focusable to another.  Just to be clear, the left/right/up/down events are not associated with any view.  Only the focus/blur/select events are specific to a view.

I am considering merging the work I did here, but it would need to be optional, as not all tvOS developers may want to pull in the GameController framework by default.

**Comment by NickFoden** [comment]
Oh fascinating. Maybe I missed a piece of setup or using incorrectly. 

With 
```javascript
"react": "17.0.1",
"react-native": "npm:react-native-tvos@0.64.2-1",
```

I was listening with the useTVEventHandler and never seeing a directional event. only focus, blur, select, etc

```javascript

const myTVEventHandler = (evt) => {
     console.log(evt.eventType)
  }

 useTVEventHandler(myTVEventHandler);
 ```

**Comment by douglowder** [comment]
In the simulator, try using the arrow keys to see the events.  Also, make sure you have more than one focusable in your view so that there is somewhere to navigate to...

**Comment by NickFoden** [comment]
Ah thank you for the information and excuse me for the spam/tangents on this thread. I  am navigating around a grid of TouchableOpacity items. Maybe I am missing a prop on each or such, have been searching for more examples of react-native-tvos.

```javascript
        <FlatList
          data={filteredVideos}
          numColumns={6}
          renderItem={({item}) => (
            <TouchableOpacity
              key={item._id}
              onPress={() => {
                setState(S => ({...S, selected: item._id}));
              }}
              onFocus={() => {
                setState(S => ({...S, preview: item._id}));
              }}>
              <View>
                <Image
                  style={styles.thumbnail}
                  source={{
                    cache: 'default',
                    uri: item.promotionalMediaUrl,
                  }}
                />
                <Text style={{color: 'white'}}>
                  {item.headline.slice(0, 15)}
                </Text>
              </View>
            </TouchableOpacity>
          )}
          keyExtractor={({_id}) => _id}
        />

**Comment by nriccar** [comment]
Hey Douglas! Will the top left and top right features be added on the main branch?

Thanks!

@andremaximino-wit how did you manage to expand the area for left and right events? Did you lower the treshold?

Do you think its feasible doing a longpress left - right listener?

**Comment by nriccar** [comment]
Sorry for double comment. 

Can you help me logging the dpadY and dpadX? I don't get too much of objective c, maybe is a really simple one!
  ![Screenshot 2021-10-29 at 17 31 51](https://user-images.githubusercontent.com/49196314/139462415-832c385f-5a27-4416-8aea-543ea2811fd6.png)

Im noticing im not being able to log the stepRight or stepRight events. I even hardcoded it:
![Screenshot 2021-10-29 at 17 44 04](https://user-images.githubusercontent.com/49196314/139464194-1c58d3e3-9d09-4a33-9459-1088739ec8e6.png)

But its not getting logged:

![Screenshot 2021-10-29 at 17 44 25](https://user-images.githubusercontent.com/49196314/139464233-6d9a9cb2-52a8-47c1-93cc-1db468a358ce.png)

Thanks for reading

**Comment by douglowder** [comment]
@nriccar After some other items are done, I will come back to this work. I have to figure out how to make this optional so that developers who don't want this feature are not required to bring in the GameController framework.

**Comment by nriccar** [comment]
Thank you Douglas.

Can we help in any way? We really need this feature.

🙌🏼

**Comment by douglowder** [comment]
@nriccar I'm now working on this again, will keep you updated here.  Work will be in `tvos-corner-click` branch.

**Comment by nriccar** [comment]
Great! Thank you Douglas!!

**Comment by douglowder** [comment]
I'm still working on surfacing corner clicks and left/right clicks of the Siri remote.

For completeness, I should point out that for video use cases, there is a specific native API for MediaPlayer... it is discussed here: https://developer.apple.com/news/?id=33cpm46r . Not having any luck yet with using MPRemoteCommandCenter and friends.

Based on the discussion in that article, it seems like this repo already supports the hardware buttons for left/right/up/down on the new Siri remote. I'm adding support for long presses on those buttons as well, and am going to get one or two of the new remotes to test with.

**Comment by nriccar** [comment]
Hey Douglas! 

Awesome. So, is it ready under tvos-corner-click branch?

**Comment by douglowder** [comment]
@nriccar the code in tvos-corner-click branch is indeed functional. I don't think it's quite ready to merge to the main repo, but it would certainly be easy to patch `node_modules/react-native` in your existing project with this: 
https://github.com/react-native-tvos/react-native-tvos/compare/tvos-v0.64.2...tvos-corner-click.diff

The current implementation requires that you add a post_install step to your project Podfile, to turn on the preprocessor macro that enables the feature in native code:

```
post_install do |installer|
  tvos_corner_clicks_post_install(installer)
end
```

Then rerun `pod install`.

The new events (currently named `stepLeft` and `stepRight`) will only show up in testing with a real device -- the GameController APIs don't do anything in the simulator.

**Comment by nriccar** [comment]
Thank you Douglas! Will check it

**Comment by douglowder** [comment]
@nriccar @andremaximino-wit @NickFoden I did some testing today with `react-native-video`, and the `Video` component does support the corner click events natively, without any need for the changes I've been testing. Specifically, if you put the video in full screen mode (see example at https://github.com/react-native-video/react-native-video/blob/master/examples/embed-and-fullscreen/index.ios.js), the player actually allows you to step forward and backward with corner clicks (and even shows a "+10s" tooltip in the progress bar when you touch the corner without clicking). FYI in case the native video player is sufficient for your use case.

**Reply by nriccar** [reply]
Thank you Douglas!

**Comment by douglowder** [comment]
Moving this to the backlog, as it is not needed for full screen videos, and could be difficult to support if there are issues with the custom gesture recognizers (especially on Fabric).
